blockchain_wasi.md

```diff
@@ -0,0 +1,200 @@
+
+## Copyright Notice
+
+  Copyright (c) 2019 Oasis Labs and the persons identified as the
+  document authors.
+
+  Code Components extracted from this document must include Simplified
+  BSD License and are provided without warranty as described in the
+  Simplified BSD License.
+
+1. Introduction
+
+  WebAssembly provides a similar set of goals (cross-platform, compactness,
+  sandboxing, performance), as have been traditionally strived for in a
+  blockchain context. Blockchain has traditionally favored determinism over
+  performance, however it is notable that WASM is also deterministic except
+  for handling of floating point NaN [WADETE]. What is needed to use
+  WebAssembly as a basis for a blockchain computation platform is access to
+  chain state.
+
+  The WASI systems interface provides a limited interface to an external
+  system for WebAssembly. 
+
+2.  Normative Language
+
+  The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+  "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+  "OPTIONAL" in this document are to be interpreted as described in
+  BCP 14 [RFC2119] [RFC8174] when, and only when, they appear in all
+  capitals, as shown here.
+
+
+3. Motivation
+
+The primary motivation of this interpretation is to allow a program
+targeting a WASI compilation target to work in a blockchain context
+with minimal if any modification, while exposing the capabilities to
+allow extension for blockchain interaction.
+
+4. Behavior of WASI in a Blockchain Context
+
+  A transaction MUST execute the WASI program, with input to the
+  transaction available on the `stdin` stream passed to the program.
+  Output of the program, as written to the `stdout` default file
+  descriptor, will be considered to be the transaction output.
+  The `proc_exit` call indicates the termination of execution, with
+  the return status code used to determine if the transaction
+  completed (0) or should be reverted (non-zero).
+
+  The `stderr` (fd 2) stream MAY be used for debugging. When running
+  a contract in debugging mode, the output to `stderr` MAY be made
+  available to the caller even if the transaction execution aborts.
+  When running in a production mode, output to `stderr` SHOULD be
+  discarded.
+
+  On-chain storage MUST be exposed as a virtual file system.
+
+  A specific namespace SHOULD be reserved within the virtual file system
+  for exposing blockchain-specific functionality.
+  This functionality SHOULD be identified as `/opt/CHAIN`, where
+  `CHAIN` MAY be either a common short name identifying the network
+  OR the network identifier adopted by the chain.
+
+  * `/opt/CHAIN/ADDRESS` SHOULD should be used as the conventional
+  location for references to other contracts. Within these directories
+  the following files SHOULD be exposed:
+  * `code` - The code at that address, if any.
+  * `sock` - A file descriptor for cross-contract calls to that address.
```

Comment:
ideally would like to be able to express the distinction between `cross contract call`, and `pure call`- where all state viewable to the delegated call is transient

Reply:
I think that the real issue is that current tooling doesn't expose wasm imports/exports without annoying `extern "C"` functions. With enough codegen, anything is possible, but it's not straightforward and, even then, calling a function doesn't imply the runtime cost of the potentially cross-shard RPC that is actually executed.

Reply:
the other way to look at this is: can i learn semantics about a remote library/service to gain safety / does the calling program have any ability to run a program with constraints?

* call this other service without the ability for it to generate logs
* call this other service without the ability for it to read from / write to persistent storage
* call this other service with an allowance of up to <x> gas

Reply:
Update: `sock` isn't the right abstraction since it doesn't have a notion of value transfer. Instead, a cross-contract call should be crated by `__wasi_blockchain_transact` which populates a FD that yields the output of the call.